data/libs/Passengers.lua

```diff
@@ -0,0 +1,183 @@
+end
+
+-- Function: CountFreeCabins
+--
+-- Return the number of unoccupied cabins present on the ship
+--
+---@param ship Ship
+---@return integer
+function Passengers.CountFreeCabins(ship)
+	return (ship["cabin_cap"] or 0) - (ship["cabin_occupied_cap"] or 0)
+end
+
+-- Function: GetOccupiedCabins
+--
+-- Return a list of currently occupied cabin equipment items
+--
+---@param ship Ship
+---@return Equipment.CabinType[]
+function Passengers.GetOccupiedCabins(ship)
+	local equipSet = ship:GetComponent("EquipSet")
+
+	return equipSet:GetInstalledWithFilter('Equipment.CabinType', function(equip)
+		return equip:GetNumPassengers() > 0
+	end)
+end
+
+-- Function: GetFreeCabins
+--
+-- Return a list of currently free passenger cabins
+--
+---@param ship Ship
+---@param numPassengers integer?
+---@return Equipment.CabinType[]
+function Passengers.GetFreeCabins(ship, numPassengers)
+	local equipSet = ship:GetComponent("EquipSet")
+
+	return equipSet:GetInstalledWithFilter('Equipment.CabinType', function(equip)
+		return equip:GetFreeBerths() >= (numPassengers or 1)
+	end)
+end
+
+-- Function: CheckEmbarked
+--
+-- Validate how many of the given list of passengers are present on this ship
+--
+---@param ship Ship
+---@param passengers Character[]
+---@return integer
+function Passengers.CheckEmbarked(ship, passengers)
+	local occupiedCabins = Passengers.GetOccupiedCabins(ship)
+	local passengerLookup = utils.map_table(passengers, function(_, passenger)
+		return passenger, true
+	end)
+
+	local count = 0
+
+	for _, cabin in ipairs(occupiedCabins) do
+		for _, passenger in ipairs(cabin.passengers) do
+			if passengerLookup[passenger] then
+				count = count + 1
+			end
+		end
+	end
+
+	return count
+end
+
+-- Function: EmbarkPassenger
+--
+-- Load the given passenger onboard the ship, optionally to a specific cabin.
+--
+---@param ship Ship
+---@param passenger Character
+---@param cabin EquipType?
+function Passengers.EmbarkPassenger(ship, passenger, cabin)
+	---@cast cabin Equipment.CabinType?
+	if not cabin then
+		cabin = Passengers.GetFreeCabins(ship)[1]
+	end
+
+	if not cabin then
+		return false
+	end
+
+	cabin:AddPassenger(passenger)
+	ship:setprop("cabin_occupied_cap", (ship["cabin_occupied_cap"] or 0) + 1)
+
+	return true
+end
+
+-- Function: DisembarkPassenger
+--
+-- Unload the given passenger from the ship. If not specified, the function
+-- will attempt to automatically determine which cabin the passenger is
+-- embarked in.
+--
+---@param ship Ship
+---@param passenger Character
+---@param cabin EquipType?
+function Passengers.DisembarkPassenger(ship, passenger, cabin)
+	---@cast cabin Equipment.CabinType?
+	if not cabin then
+		local equipSet = ship:GetComponent("EquipSet")
+
+		---@type Equipment.CabinType[]
+		local cabins = equipSet:GetInstalledWithFilter('Equipment.CabinType', function(equip)
+			return equip:HasPassenger(passenger)
+		end)
+
+		cabin = cabins[1]
+	else
+		cabin = cabin:HasPassenger(passenger) and cabin or nil
```

Comment:
Nitpick: Should this be checked first and if the passenger is not in the indicated cabin, then search the remaining cabins regardless?

Or is it intended that this should fail if the passenger is not in the intended cabin?

Reply:
If the calling function is passing an explicit cabin parameter, it is assumed the caller is already doing bookkeeping to determine which cabins specific passengers are in. If the passenger is not in the cabin the calling code is specifically asking for removal from, it is a program error.